policies/codingstyle.txt

```diff
@@ -289,12 +293,94 @@ options.
 
 In source files, separate functions with one blank line.
 
+                Chapter 6.1: Function Arguments
+
+Most API functions in OpenSSL revolve around a specific type (e.g. SSL_CTX,
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_set_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
+
+Some more complex APIs with diverse underlying implementations have TYPE
+paired with a TYPE_METHOD, where TYPE usually represents a collection of
+attributes while TYPE_METHOD represents the set of function pointers to the
+implementation. The public API remains as above, except for the constructor,
+which should look like one of these (the latter being a constructor that
+uses a factory pool referred to by some context):
+
+    TYPE *TYPE_new(TYPE_METHOD *, [args])
+    TYPE *TYPE_new(OSSL_LIB_CTX *, TYPE_METHOD *, [args])
+
+A variant for the same sort of complex APIs is also seen, with the
+corresponding type pair being TYPE_CTX and TYPE, where TYPE usually
+represents the set of function pointers to the implementation, while
+TYPE_CTX is the collection of attributes to perform the intended operation.
+For that case, the pattern should include:
+
+    TYPE_CTX *TYPE_CTX_new([args])
+    rettype TYPE_CTX_get_something(TYPE_CTX *,[args])
+    rettype TYPE_CTX_set_something(TYPE_CTX *,[args])
+    rettype TYPE_do_something(TYPE_CTX *,[args])
+
+Where present the TYPE (or TYPE_CTX) argument should always be first as shown
+in the examples above.
+
+For all of the above patterns, the remaining argument list should be put into
+order of importance. Deciding the relative importance of arguments is
+necessarily somewhat subjective, but the following guideline ordering of the
+arguments should be considered:
+
+1) Additional context arguments
+
+   If multiple contexts are necessary then they should be in order of most
+   general to most specific. In some cases (typically with an OSSL_LIB_CTX), a
+   context may be NULL to indicate a default context should be used.
+
+2) Required arguments
+
+3) Optional arguments
+
+   Here optional are arguments are meant to be used in certain circumstances
+   only and therefore may be "NULL" or empty.
+
+4) Output parameters
```

Comment:
Darn it, we may be hitting a hurdle with this one.  See https://github.com/openssl/openssl/pull/12750#discussion_r489713752

Did anyone look at current API to see what pattern we already have for this?  It's going to be painful if, every time that someone looks at already existing to make similar enough functions, just to having someone brandish the coding guidelines with a "but but but"...  in the end, we will need to consider the element of least surprise for the user, and one set of EVP_PKEY functions deviating too much from already established patterns within EVP risks being more painful than we're comfortable with.

FYI, I have not looked around more than what I've mentioned in the comment that I linked to...